metabase/metabase
# RFC 70 - Automate sorting imports

https://www.notion.so/metabase/70-Automate-sorting-imports-74133d1e110e4605b079d153baf57953

### Description

- Due to a huge diff the only way to review this is to go over [commits](https://github.com/metabase/metabase/pull/33883/commits) one by one
- There are 2 commits containing automatic fixes ([Run "yarn eslint-fix" 5 times](https://github.com/metabase/metabase/pull/33883/commits/a95906a9b3bf5af7da92ccbc1c2e446e7fcc2bba) & [Run "yarn prettier"](https://github.com/metabase/metabase/pull/33883/commits/004d139aef2dc500b9a6294adce8b526a094e350)), all other were manual or quasi-manual
    - I had to run `yarn eslint-fix` multiple times, as it was not able to fix all issues in a single run - with every subsequent run there were fewer and fewer rule violations
- `import/order` rule does not play well with comments or code in between import statements, so I had to do some adjustments before running `yarn eslint-fix`:
  - (75aa0de96bad2f9aeb0d6fdc714e12be44700979, 898d16ed8cc9061c453a20700b45815cb8578ec9) moved comments into the same line as imports
  - (d6b4e6d4395e6f4c40d64d57c5c8fd3d50de52a1) moved imports into a single block (where possible)
  - (1ae842d36e0bf967ace351fcd3984801c2e2d07a) moved or removed informative comments between import statements
  - (e1e072db4eb1df81c9fee182d4560f8df5a6c00d) removed imports-grouping comments (now superseded by `import/order` rule)
  - (071432d50668e907a3835b7e52b3ef53dbad4429) disabled `import/order` rule in files where apparently the order of imports matters
- After `yarn eslint-fix` there were still some files that eslint complained about, so I manually fixed them (b2e4a5420f037acbe3f5afef53ddd751b3aa5190)
- After all this there was 1 circular dependency issue that came up in unit tests - fixed in [this commit](https://github.com/metabase/metabase/pull/33883/commits/b31ec8d03a8e50e25b3ecc04cee37eb9739f9254)

This change should not be backported. I tried to work on this as soon as we cut the `release-x.49.x` branch.

### How to verify

- CI checks should pass (with exception of unrelated flaky tests mentioned below)
- App should work exactly the same
- Verify commits one by one
    - it's ok to skip the commits that merge `master` into this branch
    - it's beyond reasonable to manually review [Run "yarn eslint-fix" 5 times](https://github.com/metabase/metabase/pull/33883/commits/a95906a9b3bf5af7da92ccbc1c2e446e7fcc2bba) & [Run "yarn prettier"](https://github.com/metabase/metabase/pull/33883/commits/004d139aef2dc500b9a6294adce8b526a094e350) commits

### Living with `import/order`

Since `import/order` rule does not play well with comments (or code) in between import statements, it's advised to put comments on the same line as imports, e.g.:

:x: do not do this:
```js
// this is a comment
import something from "./somewhere";
```

:heavy_check_mark: do this instead:
```js
import something from "./somewhere"; // this is a comment
```

Otherwise `import/order` will not be able to correctly auto-fix the rule violations.

### Failing tests

Failing tests are unrelated to this PR:
- `e2e/test/scenarios/permissions/sandboxes.cy.spec.js` ([e2e-tests-flaky-ee](https://github.com/metabase/metabase/actions/runs/7928802073/job/21649260820?pr=33883#logs)) is a [known flake](https://metaboat.slack.com/archives/C5XHN8GLW/p1708075278201149) in `master`
- `QueryBuilder.unsaved-changes-warning.unit.spec.tsx` ([fe-tests-unit](https://github.com/metabase/metabase/actions/runs/7930239252/job/21653615068?pr=33883)) is a [known flake](https://metaboat.slack.com/archives/C5XHN8GLW/p1708086419359699) in `master`

### Solving conflicts after this is merged

How to resolve conflicts:
1. Make sure your local `master` is up to date with remote
2. Checkout your feature branch
3. `git merge master`
4. run `yarn` to install the latest version of `eslint-plugin-import`
5. reload your IDE to make sure new ESLint configuration kicked in
6. for every file that has conflicts **around `import` statements**:
    1. resolve conflicts using **your changes**
    2. run `eslint --fix` on that file (can be done via IDE)
